doc/developer/logging.rst

```diff
@@ -239,3 +242,180 @@ Examples:
+          struct zlog_target zt;
+          int syslog_facility;
+      };
+
+      static void zlog_syslog(struct zlog_target *zt, struct zlog_msg *msgs[], size_t nmsgs)
+      {
+          struct zlt_syslog *zte = container_of(zt, struct zlt_syslog, zt);
+          size_t i;
+
+          for (i = 0; i < nmsgs; i++)
+              if (zlog_msg_prio(msgs[i]) <= zt->prio_min)
+                  syslog(zlog_msg_prio(msgs[i]) | zte->syslog_facility, "%s",
+                         zlog_msg_text(msgs[i], NULL));
+      }
+
+
+.. c:function:: struct zlog_target *zlog_target_clone(struct memtype *mt, struct zlog_target *oldzt, size_t size)
+
+   Allocates a logging target struct.  Note that the ``oldzt`` argument may be
+   ``NULL`` to allocate a "from scratch".  If ``oldzt`` is not ``NULL``, the
+   generic bits in :c:type:`struct zlog_target` are copied.  **Target specific
+   bits are not copied.**
+
+.. c:function:: struct zlog_target *zlog_target_replace(struct zlog_target *oldzt, struct zlog_target *newzt)
+
+   Adds, replaces or deletes a logging target (either ``oldzt`` or ``newzt`` may be ``NULL``.)
+
+   Returns ``oldzt`` for freeing.  The target remains possibly in use by
+   other threads until the RCU cycle ends.  This implies you cannot release
+   resources (e.g. memory, file descriptors) immediately.
+
+   The replace operation is not atomic; for a brief period it is possible that
+   messages are delivered on both ``oldzt`` and ``newzt``.
+
+   .. warning::
+
+      ``oldzt`` must remain **functional** until the RCU cycle ends.
+
+.. c:function:: void zlog_target_free(struct memtype *mt, struct zlog_target *zt)
+
+   Counterpart to :c:func:`zlog_target_clone()`, frees a target (using RCU.)
+
+.. c:member:: void (*zlog_target.logfn)(struct zlog_target *zt, struct zlog_msg *msgs[], size_t nmsg)
+
+   Called on a target to deliver "normal" logging messages.  ``msgs`` is an
+   array of opaque structs containing the actual message.  Use ``zlog_msg_*``
+   functions to access message data (this is done to allow some optimizations,
+   e.g.  lazy formatting the message text and timestamp as needed.)
+
+   .. note::
+
+      ``logfn()`` must check each individual message's priority value against
+      the configured ``prio_min``.  While the ``prio_min`` field is common to
+      all targets and used by the core logging code to early-drop unneeded log
+      messages, the array is **not** filtered for each ``logfn()`` call.
+
+.. c:member:: void (*zlog_target.logfn_sigsafe)(struct zlog_target *zt, const char *text, size_t len)
+
+   Called to deliver "exception" logging messages (i.e. SEGV messages.)
+   Must be Async-Signal-Safe (may not allocate memory or call "complicated"
+   libc functions.)  May be ``NULL`` if the log target cannot handle this.
+
+Standard targets
+^^^^^^^^^^^^^^^^
+
+:file:`lib/zlog_targets.c` provides the standard file / fd / syslog targets.
+The syslog target is single-instance while file / fd targets can be
```

Comment:
Log targets can access the raw format string, and instead of formatting you could serialize/save the printf arguments from `zlog_msg->args`.

(I've kept `zlog_msg` private to the logging code for now, but we can add access functions or make it public as needed.)